server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringRareTermsAggregator.java

```diff
@@ -0,0 +1,213 @@
+import org.apache.lucene.util.CollectionUtil;
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.util.BytesRefHash;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * An aggregator that finds "rare" string values (e.g. terms agg that orders ascending)
+ */
+public class StringRareTermsAggregator extends AbstractRareTermsAggregator<ValuesSource.Bytes, IncludeExclude.StringFilter> {
+    // TODO review question: is there equivalent to LongObjectPagedHashMap like used in LongRareTerms?
+    protected ObjectLongHashMap<BytesRef> map;
+    protected BytesRefHash bucketOrds;
+    private LeafBucketCollector subCollectors;
+
+    // TODO review question: What to set this at?
+    /**
+     Sets the number of "removed" values to accumulate before we purge ords
+     via the MergingBucketCollector's mergeBuckets() method
+     */
+    private final long GC_THRESHOLD = 10;
+
+    StringRareTermsAggregator(String name, AggregatorFactories factories, ValuesSource.Bytes valuesSource,
+                                     DocValueFormat format,  IncludeExclude.StringFilter stringFilter,
+                                     SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+                                     Map<String, Object> metaData, long maxDocCount) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData, maxDocCount, format, valuesSource, stringFilter);
+        this.map = new ObjectLongHashMap<>();
+        this.bucketOrds = new BytesRefHash(1, context.bigArrays());
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        final SortedBinaryDocValues values = valuesSource.bytesValues(ctx);
+        if (subCollectors == null) {
+            subCollectors = sub;
+        }
+        return new LeafBucketCollectorBase(sub, values) {
+            final BytesRefBuilder previous = new BytesRefBuilder();
+            private long numDeleted = 0;
+
+            @Override
+            public void collect(int docId, long bucket) throws IOException {
```

Comment:
Same comments apply as from `LongRareTermsAggregator` above. Also since this logical is almost the same in three places does it make sense to extract it to something common so we can fix it in one place and apply it to all implementations?

Reply:
I tried hard to refactor collect() and gcDeletedEntries() into one place... and it's just not possible.  There are too many differences between longs and BytesRef.  Map get/set, ordinals, hashing, doc values, etc are all different and there aren't any shared types that allow it to be resolve easily :(